### ISSUE TITLE
Flutter performance profiler is causing massive jank when profiling

### ISSUE THREAD

**Issue by matthew-carroll** [flutter/flutter#127276]
I'm trying to use the Flutter profiler to locate sources of jank in my UI. Unfortunately, the Flutter profiler app, itself, seems to be causing almost all of the jank that occurs when I'm using it.

Here's the reported frame times when I haven't opened the profiler app

https://github.com/flutter/flutter/assets/7259036/659f2fbd-53da-48ff-b8f8-94418141226f

Here's the reported frame times when I'm using the profiler app

https://github.com/flutter/flutter/assets/7259036/2abced08-e64c-407f-89db-0cc3fe0ccb09

Here are screenshots of an arbitrary jank frame from the profiler app

<img width="1456" alt="Screen Shot 2023-05-20 at 1 58 10 PM" src="https://github.com/flutter/flutter/assets/7259036/8f393eee-a546-44b7-a18d-d771a1a9e8d5">
<img width="1453" alt="Screen Shot 2023-05-20 at 1 58 20 PM" src="https://github.com/flutter/flutter/assets/7259036/0648fb29-5151-4020-af2b-cc87a39c9c7f">

Flutter version:
```
Flutter 3.10.0 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 84a1e904f4 (11 days ago) • 2023-05-09 07:41:44 -0700
Engine • revision d44b5a94c9
Tools • Dart 3.0.0 • DevTools 2.23.1
```

**Comment by matthew-carroll** [comment]
CC @slightfoot

**Comment by matthew-carroll** [comment]
CC @kenzieschmoll (in case the perf characteristics are under your purview)

**Comment by darshankawar** [comment]
Thanks for the report @matthew-carroll 
Any code sample or steps to replicate would be helpful to properly analyze the reported behavior.

**Comment by kenzieschmoll** [comment]
The top video you linked from IntelliJ is actually a bit deceiving. That chart only shows the times for the UI portion of each Flutter frame, whereas DevTools shows the times for both the UI and the Raster portions of each frame. Looking at the chart in DevTools, it looks like all your jank is actually raster thread jank, which would make sense why the frame times show up as performant in IntelliJ (which doesn't know anything about the raster thread activity).

In DevTools, if you click on a frame to zoom to its timeline events, you should see the UI thread activity is significantly shorter than the raster thread activity where the jank is.

You can try the [Raster Stats tool](https://docs.flutter.dev/tools/devtools/performance#raster-stats-tab) to try to identify the source of your raster thread jank (though this tool is designed to work best with impeller, as the stats will be inconsistent when using SKIA due to shader warmups).

**Comment by matthew-carroll** [comment]
@kenzieschmoll we can't use Impeller because we're deploying to Android. What should I use to inspect Skia raster time in a consistent way?

**Comment by kenzieschmoll** [comment]
You can look at the timeline events on the raster thread to inspect raster time for Skia. You can also try commenting out parts of the widget tree until the raster time you see in the performance page improves. The good news is you can debug raster time even in debug mode so you can use hot reload to make this iteration process quicker. If you can do this to the point where you find the expensive widget(s}, that should give you a decent start on finding the problem. 

Also consider using the "More debugging options" menu to disable certain rendering layers to see if that helps: https://docs.flutter.dev/tools/devtools/performance#more-debugging-options

CC @dnfield in case you have other suggestions.

**Comment by matthew-carroll** [comment]
Using the raster stats, I identified the culprit. It must be some kind of Flutter engine inefficiency. We have a `CustomPaint` plot to measure velocity for one of widgets. We draw plot points as circles, and lines between them. We're only plotting a max of 300 points, but that plot is creating all the raster jank. Even more interesting is that the `drawLine()` in particular is the majority of that jank. Anyway, I guess I need to file that as another bug.

But before leaving this topic, I noticed that the raster jank is significantly worse in Profile mode as compared to Debug mode.

Here's the Debug mode timeline:
<img width="1446" alt="Screen Shot 2023-05-22 at 2 12 19 PM" src="https://github.com/flutter/flutter/assets/7259036/6afe4dc7-afca-4f43-b70f-90d795c71e6f">

Here's the Profile mode timeline:
<img width="1453" alt="Screen Shot 2023-05-22 at 2 13 12 PM" src="https://github.com/flutter/flutter/assets/7259036/cbb6e388-f1d4-49c4-9958-ae7979a94b4e">

Any idea why this difference exists?

This is a very counterintuitive performance result. Typically, I don't even look at performance characteristics in Debug mode. There's even that scary warning in Dev Tools when we're in Debug mode. So typically I jump right to Profile mode, with the understanding that Profile mode is as close as we get to Release mode in terms of performance, but still with the necessary monitoring tools. I don't know what to make of a situation like this. Would Release mode show the characteristics of Profile mode in this case, or more like Debug mode? It's unclear what a Flutter dev is expected to do in a situation like this.

**Comment by matthew-carroll** [comment]
@kenzieschmoll is there a way to generate a raster layer snapshot for a single frame? For example, to diagnose a situation like this:

<img width="642" alt="Screen Shot 2023-05-22 at 2 25 02 PM" src="https://github.com/flutter/flutter/assets/7259036/5af41d8d-835f-4f84-9121-f21514ebcdf4">

**Comment by kenzieschmoll** [comment]
> is there a way to generate a raster layer snapshot for a single frame? For example, to diagnose a situation like this:

There is not currently a way to do this, no. This functionality is on our radar though: https://github.com/flutter/devtools/issues/4517. CC @iskakaushik

**Comment by kenzieschmoll** [comment]
Tagging @dnfield and @jonahwilliams for the debug vs profile mode discrepancy [above](https://github.com/flutter/flutter/issues/127276#issuecomment-1558024488).

**Comment by dnfield** [comment]
It's hard for me to tell from that screenshot what the numbers are. It looks like debug may have had a few faster and a few much slower frames, whereas profile had a more even distribution of fast and slow frames (but none quite as fast or slow).

If this test was done manually, it's entirely possible that the test was run slightly differently between different runs as well.

**Comment by matthew-carroll** [comment]
@dnfield these perf reports come from the same experience that yielded the minimal repo in the other ticket I opened today: https://github.com/flutter/flutter/issues/127356

You're welcome to run that code any way that you think might shed light on issues or unexpected profiling results.

**Comment by kenzieschmoll** [comment]
In DevTools, we listen for 'Flutter.Frame' events, which have a field for 'build' time and for 'raster' time. These are the values we use for the DevTools frames chart. If either of these times is greater than the budget per frame (~11 ms for a 90 FPS device like in your screenshot), we show jank.

In IntelliJ, it looks like we use the 'elapsed' field from the 'Flutter.Frame' events to determine jank: https://github.com/flutter/flutter-intellij/blob/fe38c1f0d4885ced04c1c22defd55c6d42835df2/flutter-idea/src/io/flutter/vmService/FlutterFramesMonitor.java#L53. This explains the discrepancy, and would indicate that your performance issue may actually be with latency between frames since the 'build' and 'raster' times are within the budget for 90 FPS. There are a couple open issues on the DevTools issue tracker related to jank vs latency and how we can help users debug latency issues.
- https://github.com/flutter/devtools/issues/451
- https://github.com/flutter/devtools/issues/4479

CC @jonahwilliams @chinmaygarde @dnfield

**Comment by matthew-carroll** [comment]
@kenzieschmoll I'm curious about build vs raster time. It sounds like DevTools treat those as overlapping times, i.e., the next frame's build runs during the previous frame's raster. I assume that's why you're saying 11ms for either build OR raster, but not build AND raster.

My question is, are we guaranteed that these overlap? Or is that just the nominal case? I believe I've seen many situations where they don't seem to overlap. Where the next frame's build is shown on the timeline AFTER the previous frame's raster.

If that's just a visualization discrepancy, then so be it. But, if there are actually scenarios where build and raster do not overlap, and instead end up back-to-back, I think it's very important to reveal that information, and mark those frames as janky if build and raster, combined, exceed the 11ms for 90FPS, or 16ms for 60FPS.

**Comment by kenzieschmoll** [comment]
@dnfield @chinmaygarde @jonahwilliams do you have opinions on how we should be calculating frame times and determining jank?

**Comment by kenzieschmoll** [comment]
> If a single build + raster takes > vsync period, then you will likely hit jank

How is "vsync period" defined by timeline events?

> The real measure should be whether the raster frame managed to submit before the next vsync signal

And if the raster portion of a frame finishes before the vysnc signal, then this means there _is_ or _is not_ jank? Also what is the "vsync signal" in terms of timeline events?

**Comment by a-siva** [comment]
See go/perfetto-speed-size-comparison for some discussions on how collecting profiling data can be improved

**Comment by zanderso** [comment]
@jmquilez it looks like you are experiencing a different problem than the one this issue is about. Please file a new issue with these details. Thanks!

**Comment by matthew-carroll** [comment]
@jmquilez maybe you're looking for this issue? https://github.com/flutter/flutter/issues/129150

**Comment by jmquilez** [comment]
Hello @zanderso @matthew-carroll 
I decided to comment here because of what @matthew-carroll mentioned in this comment https://github.com/flutter/flutter/issues/127276#issuecomment-1569222132 about android studio showing actual janky frames but devtools omitting them; I have the same problem, but with the difference that in my case the app context is a listview of videos and the jank is manifested in the form of "jumps" as I scroll.

> @jmquilez maybe you're looking for this issue? https://github.com/flutter/flutter/issues/129150

I believe it'll might be related in the sense that you mention devtools are not showing janky frames, however as I can see the context is different: in #129150 the problem happens in Android tablets and only when zooming / panning.

Anyway if you believe this will be better discussed in a new issue, I'll make sure to open it.

Thanks for your help and time :)

**Comment by derekxu16** [comment]
We are planning on adding an option to the `flutter` tool that will allow you to write a timeline trace to a file, that can then be loaded into [Perfetto's trace viewer](https://ui.perfetto.dev/). This will allow you to avoid any overhead from DevTools.

**Comment by matthew-carroll** [comment]
That is probably a technical workaround, but it should be well understood that saving to a file and loading that file into another tool is a major step back in terms of DevX. 

I've mentioned cycle friction in other tooling tickets. I'll repeat it here. When debugging performance issues, developers often have to dig through deep code, typically code they don't own, and correlate that code with chart movement. Developers also typically need to use guesswork to see what results in better charted performance. Asking developers to go through dozens or hundreds of these debugging cycles, and do so by downloading a file and opening it in another tool, will be hugely frustrating. I think the team should look for paths that allow developers to debug real performance issues in as close to a realtime environment as possible.

**Comment by Hixie** [comment]
It looks like the original issue filed here ("Flutter performance profiler is causing massive jank when profiling") was deemed invalid, and the subsequent discussion is about various issues arising from the app under test. It's not clear that this specific issue is actionable by the Flutter project at this time.

@matthew-carroll Would it make sense to close this issue and file new ones on the specific problems with Flutter that you've uncovered?

**Comment by matthew-carroll** [comment]
@Hixie I'm not sure this issue is "invalid" - it mostly seems under-investigated. It might be valid, or invalid, but that's unclear because the issues raised WRT the performance profiler were never objectively resolved. If you don't want any further investigation into those issues, then you can close this ticket.

The app-specific behavior that I mentioned in one comment has already been filed elsewhere.

Most of the second half of this conversation was between Flutter and Dart team members. They seemed to be discussing some deep reporting details. If those details are worthy of new tickets, then I think those people should file the issues, because I don't know anything about that internal reporting system.

**Comment by a-siva** [comment]
@matthew-carroll can you summarize here the issues you have with the performance profiler

From scanning the comments in this issue I could gather the following
- raster jank in profile mode is worse than debug mode even after --trace-skia was removed, I believe the reason was explained in a follow up comment
- you consider the workflow of saving the timeline information to a file and then viewing it offline in devtools not desirable (this is being proposed to reduce the overhead of the tool which would be very relevant in some very sensitive performance jank situations)

If there are other issues can you please list them here.

**Comment by matthew-carroll** [comment]
@a-siva both of those are correct - however, you mentioned that excessive profile jank was "explained". I would suggest that an explanation is not sufficient. It's a legitimate problem in need of a solution, which leads to your 2nd point.

---

An additional issue is the discrepancy between Android Studio frame time reporting and the Dev Tools frame time reporting. In this thread I've shown that these two frame reports can be wrong in both directions:

- The Android Studio frame time reporter shows good frame times, while the Dev Tools shows that frame times are much too high, causing jank.
- The Android Studio frame time reporter shows that frames are too long and causing jank, but the Dev Tools shows the frame times as performant.

These metrics should be the same, and they need to be accurate. Neither one should lie to developers by saying that frames are performant when they're actually janky.

---

I mentioned that it would be helpful to inspect a raster layer snapshot for a single frame, because it's fairly common to end up with one or two very long frame times, but the rest of the frames are OK. Kenzie said that's covered here: https://github.com/flutter/devtools/issues/4517

---

Then there was some discussion among team developers about internal timing communication. I haven't looked deeply into those conversations.

**Comment by Hixie** [comment]
@matthew-carroll It sounds like from the discussion in the last couple of comments there are multiple actual specific actionable issues here. Would you mind filing a separate issue for each one?
